File: benchmarks/python/nqueens.py
#!/usr/bin/env python3

from __future__ import annotations


def safe(queen: int, diag: int, xs) -> bool:
    if xs is None:
        return True
    q, tail = xs
    return queen != q and queen != q + diag and queen != q - diag and safe(queen, diag + 1, tail)


def place(n: int, row: int, soln) -> int:
    if row == 0:
        return 1
    return try_col(n, row, soln, n)


def try_col(n: int, row: int, soln, col: int) -> int:
    if col <= 0:
        return 0
    if safe(col, 1, soln):
        return place(n, row - 1, (col, soln)) + try_col(n, row, soln, col - 1)
    return try_col(n, row, soln, col - 1)


def queens(n: int) -> int:
    return place(n, n, None)


def main() -> int:
    print(queens(13))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: benchmarks/haskell/nqueens.hs
data Solution = Nil | Cons !Int !Solution

safe queen diag xs
  = case xs of
      Nil -> True
      Cons q t -> queen /= q && queen /= q + diag && queen /= q - diag && safe queen (diag + 1) t

place n 0 soln
  = 1
place n row soln
  = try n
  where
    try col
      = if col <= 0
         then 0
         else if safe col 1 soln
              then place n (row - 1) (Cons col soln) + try (col - 1)
              else try (col - 1)

queens n
  = place n n Nil

main
  = print (queens 13)
